Preferred option

huntc opened the pull request:
This DeferredResult's design is strongly influenced by the JDK 8's java.util.stream.Stream class and Scala's scala.concurrent.Future.

If the JDK 8 CompletableFuture were to implement this interface then it would be consistent with JDK 8 Stream.

Another significant advantage of this API is that there are no opportunities for blocking operations.

huntc committed:
Re-adding

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+    <A> Iterable<DeferredResult<T>> traverse(Executor executor,
+                                             Iterable<A> collection,
+                                             Function<? super A, DeferredResult<? extends T>>... suppliers);
+
```
To be honest I still don't understand these supposedly "static" methods. They should not be in the interface if they cannot be static.

huntc replied:
I've now moved to JDK 8 as per @jroper's suggestion. They will now be declared as static within the interface given JDK 8's ability to do this.

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+    boolean isCompleted();
+
+    /**
+     * The value of this `Future`.
```
The `current` value of this `Future`.

huntc replied:
fixed.

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+
+    DeferredResult<Long> mapToLong(ToLongFunction<? super T> mapper);
+
+    DeferredResult<Long> mapToLong(ToLongFunction<? super T> mapper, Executor executor);
```
What are these: mapToX? They should be dropped IMO

jroper replied:
I _think_ they are not needed.  The Java 8 Stream API has mapToX, but that's because they have XStream, and presumably this is for performance, so you can have for example IntStream without boxing each int.  Since we don't have XDeferredResult, we don't need mapToX.  I also don't think we need XDeferredResult, since a future is quite heavy weight compared to a single element in a stream, so boxing primitives in DeferredResult shouldn't really hurt.

huntc replied:
I will drop them then.

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+
+    DeferredResult<Long> flatMapToLong(Function<? super T, ? extends DeferredResult<Long>> mapper);
+
+    DeferredResult<Long> flatMapToLong(Function<? super T, ? extends DeferredResult<Long>> mapper, Executor executor);
```
flatMapToX? they should be dropped

huntc replied:
Dropped

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+    <R> R collect(Supplier<R> resultFactory,
+                  BiConsumer<R, ? super T> accumulator,
+                  BiConsumer<R, R> combiner,
+                  Executor executor);
```
This doesn't make sense IMO, not only is it not isomorphic to Scala's collect, but who put it in?

huntc replied:
I thought that it was isomorphic to Scala's Future.collect... The signature actually comes from Stream.collect.

Can you please offer an alternate signature or corrections to the above.

viktorklang replied:
```
<R> DeferredResult<R> collect(Function<? super T, Optional<? extends R>> collector)
```

viktorklang replied:
Check `collect` here: http://www.scala-lang.org/api/current/index.html#scala.concurrent.Future

huntc replied:
Did you look at the Stream interface and the rationale for why they did what they did? I'm concerned about becoming inconsistent with Stream.collect

viktorklang replied:
I'm just saying that this version of `collect` and the Scala `collect` method are completely different.
So I vote for dropping it from the DeferredResult interface if it doesn't do the same thing as the s.c.Future version

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+     */
+    <R extends T> DeferredResult<R> fallbackTo(Function<? super T, ? extends R> fallback);
+
+    // No mapTo method as ClassTags are not available in Java.
```
Why is this comment needed?

huntc replied:
Gone

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,304 @@
+
+    <R> DeferredResult<R> recoverWith(Function<? super Throwable, DeferredResult<? extends R>> recovery, Executor executor);
+
+    // No zip method as there are no tuples in Java.
```
Why is this comment needed?

huntc replied:
More for you guys - I lifted this interface mostly from Scala's Future. 

I shall remove the comment now that you've seen it.

huntc committed:
Mostly applied JDK 8 features

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,346 @@
+ * A DeferredResult is a Future that can be observed. Actions are provided
+ * to execute when the results of operations are available.
+ * </p>
+ * DeferredResults are similar to Futures except their results cannot
```
I think this line is potentially confusing to the reader. I'd just scrap Future here and tell the reader that a DeferredResult represents a result that is not guaranteed to be immediately available.

All uses of the word "Future" below this line should be removed, we don't want to confuse the reader into thinking we're talking about j.u.c.Future.

huntc replied:
I've attempted to clear this up and have also removed references to futures. Please review as the transformations were not simply find/replace.

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,346 @@
+    boolean isCompleted();
+
+    /**
+     * The value of this current `Future`.
```
What does it return if the result is an exception?

viktorklang commented:
What I'd like to see here now in this PR is a retrofitted version of CompletableFuture implementing DeferredResult, to show that it works.

huntc committed:
Aligned package name with the JDK

huntc committed:
Tidied up the javadoc.

huntc commented on src/main/java/java/util/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,344 @@
+     * the reduced value is a mutable value holder, such as an ArrayList, and elements are incorporated by
+     * updating the state of the result, rather than by replacing the result.
+     */
+    default <R> R collect(Supplier<R> resultFactory,
```
I would like to write a unit test for this to see how it looks in the context of JDK 8. Apparently other parts of the JDK lend themselves to being used by here e.g. StringBuilder.

viktorklang replied:
Sounds like their "collect" is more like fold

huntc committed:
Work in progress re. re-working CompletableFuture implementing the preferred DeferredResult. More work to be done yet.

huntc committed:
Further refinement around the value method, which also eliminates the need for getNow

huntc committed:
Corrected the meaning of collect and recover so that they are semantically aligned with the Scala Future.

huntc committed:
Completed the failed() instance method, and aligned the creational methods with Scala's Future equivalents.

huntc committed:
The introduction of a test - it can't run here though given security exceptions. However it does help flesh out the api.

huntc committed:
Worked around UNSAFE issues (thanks Viktor!) and got the basic test working

huntc committed:
1. factory methods should return the type they are declared in
2. static factory methods should appear on DeferredResult and delegate to CompletableFuture, returning DeferredResult
3. other static methods on CompletableFuture should have a DeferredResult counterpart that compose DeferredResult objects as CompletableFutures and then call the corresponding CompletableFuture static method.
4. Don't cast DeferredResult to CompletableFuture - compose instead.

huntc committed:
Avoid passing java to scalar as it isn't yet entirely up to scratch on jdk8

huntc committed:
Some tidying and a bug fix

huntc committed:
Removed unnecessary creational methods

huntc committed:
Removed another creational method from the interface

huntc committed:
Return types should be CompletableFuture within CompletableFuture by convention

huntc commented:
No longer relevant
